cluster_generator/utils.py

```diff
@@ -1,43 +1,228 @@
+    else sys.stderr
+)
 cgLogger = logging.getLogger("cluster_generator")
 
-ufstring = "%(name)-3s : [%(levelname)-9s] %(asctime)s %(message)s"
-cfstring = "%(name)-3s : [%(levelname)-18s] %(asctime)s %(message)s"
+cg_sh = logging.StreamHandler(stream=stream)
 
-cg_sh = logging.StreamHandler()
 # create formatter and add it to the handlers
-formatter = logging.Formatter(ufstring)
+formatter = logging.Formatter(cgparams["system"]["logging"]["main"]["format"])
 cg_sh.setFormatter(formatter)
 # add the handler to the logger
 cgLogger.addHandler(cg_sh)
-cgLogger.setLevel('INFO')
+cgLogger.setLevel(cgparams["system"]["logging"]["main"]["level"])
 cgLogger.propagate = False
 
 mylog = cgLogger
 
+# -- Setting up the developer debugger -- #
+devLogger = logging.getLogger("development_logger")
+
+if cgparams["system"]["logging"]["developer"][
+    "enabled"
+]:  # --> We do want to use the development logger.
+    # -- checking if the user has specified a directory -- #
+    if cgparams["system"]["logging"]["developer"]["output_directory"] is not None:
+        from datetime import datetime
+
+        dv_fh = logging.FileHandler(
+            os.path.join(
+                cgparams["system"]["logging"]["developer"]["output_directory"],
+                f"{datetime.now().strftime('%m-%d-%y_%H-%M-%S')}.log",
+            )
+        )
+
+        # adding the formatter
+        dv_formatter = logging.Formatter(
+            cgparams["system"]["logging"]["main"]["format"]
+        )
+
+        dv_fh.setFormatter(dv_formatter)
+        devLogger.addHandler(dv_fh)
+        devLogger.setLevel("DEBUG")
+        devLogger.propagate = False
+
+    else:
+        mylog.warning(
+            "User enabled development logger but did not specify output directory. Dev logger will not be used."
+        )
+else:
+    devLogger.propagate = False
+    devLogger.disabled = True
+
+
 mp = (pc.mp).to("Msun")
 G = (pc.G).to("kpc**3/Msun/Myr**2")
 kboltz = (pc.kboltz).to("Msun*kpc**2/Myr**2/K")
-kpc_to_cm = (1.0*kpc).to_value("cm")
+kpc_to_cm = (1.0 * kpc).to_value("cm")
+
+X_H = cgparams["physics"]["hydrogen_abundance"]
+mu = 1.0 / (2.0 * X_H + 0.75 * (1.0 - X_H))
+mue = 1.0 / (X_H + 0.5 * (1.0 - X_H))
+
+# -- Utility functions -- #
+_truncator_function = lambda a, r, x: 1 / (1 + (x / r) ** a)
+
+
+class TimeoutException(Exception):
+    def __init__(self, msg="", func=None, max_time=None):
+        self.msg = f"{msg} -- {str(func)} -- max_time={max_time} s"
+
+
+def _daemon_process_runner(*args, **kwargs):
+    # Runs the function specified in the kwargs in a daemon process #
+
+    send_end = kwargs.pop("__send_end")
+    function = kwargs.pop("__function")
+
+    try:
+        result = function(*args, **kwargs)
+    except Exception as e:
+        send_end.send(e)
+        return
 
-X_H = 0.76
-mu = 1.0/(2.0*X_H + 0.75*(1.0-X_H))
-mue = 1.0/(X_H+0.5*(1.0-X_H))
+    send_end.send(result)
+
+
+def time_limit(function, max_execution_time, *args, **kwargs):
```

Comment:
@Eliza-Diggins could you give a brief description what this function is for (seems straightforward, but just in case), and what places in the code you are using it?

Reply:
Added a docstring here. In some situations with MONDian clusters, I the potential has to be solved for using a non-linear solver and it occasionally ran into long runtime issues. I wrote this to provide a maximum time limit on that and (more importantly), indicate to the user that something was actually happening! 

I'm hoping when I get the gravity PR prepared and cleaned up I'll come up with a better approach and get rid of this.